compiler/optimizations/inferConstRefs.cpp

```diff
@@ -0,0 +1,377 @@
+  RefInfo* info = infoMap[sym];
+
+  // 'info' may be null if the argument is never used.
+  if (info == NULL || info->finalized) {
+    return wasConstRef;
+  }
+
+  bool retval = true;
+
+  while (!info->todo.empty() && retval) {
+    SymExpr* use = info->todo.front();
+
+    // By popping, recursive calls won't try to deal with the same SymExpr.
+    info->todo.pop_front();
+
+    CallExpr* call = toCallExpr(use->parentExpr);
+    INT_ASSERT(call);
+
+    CallExpr* parent = toCallExpr(call->parentExpr);
+
+    if (call->isResolved()) {
+      ArgSymbol* form = actual_to_formal(use);
+      if (!inferConstRef(form)) {
+        retval = false;
+      }
+    }
+    else if (parent && isMoveOrAssign(parent)) {
+      if (!handleRHS(parent, call, use)) {
+        retval = false;
+      }
+    }
+    else if (call->isPrimitive(PRIM_MOVE)) {
+      if (use == call->get(1) && !call->get(2)->isRef()) {
+        retval = false;
+      } else {
+        // 'use' is the RHS of a MOVE
+        if (call->get(1)->isRef()) {
+          SymExpr* se = toSymExpr(call->get(1));
+          INT_ASSERT(se);
+          if (!inferConstRef(se->var)) {
+            retval = false;
+          }
+        } // otherwise, it's a deref-move
+      }
+    }
+    else if (call->isPrimitive(PRIM_ASSIGN)) {
+      if (use == call->get(1)) {
+        retval = false;
+      }
+    }
+    else if (call->isPrimitive(PRIM_SET_MEMBER) ||
+             call->isPrimitive(PRIM_SET_SVEC_MEMBER)) {
+      retval = false;
+    } else {
+      // To be safe, exit the loop with 'false' if we're unsure of how to
+      // handle a primitive.
+      retval = false;
+    }
+  }
+
+  // Should we 'correct' the ref if retval is false?
+  if (retval) {
+    if (ArgSymbol* arg = toArgSymbol(sym)) {
+      arg->intent = INTENT_CONST_REF;
+    } else {
+      INT_ASSERT(isVarSymbol(sym));
+      sym->qual = QUAL_CONST_REF;
+    }
+  }
+
+  // TODO: assert that it's not already finalized?
+  info->finalized = true;
+
+  return retval;
+}
+
+//
+// For each reference in the AST, determine whether or not it can be a
+// const-ref. If it can, this function will change the intent or qual of the
+// Symbol.
+//
+// This function can be called multiple times.
+//
+void inferConstRefs() {
+  // List of Symbols to infer const-ness of
+  std::vector<Symbol*> todo;
+
+  // Build a map from Symbols to RefInfo. Like buildDefUseMaps, except we don't
+  // care about the distinction between a def and a use. We just want all
+  // SymExprs for a Symbol.
+  forv_Vec(SymExpr, se, gSymExprs) {
+    if (!(isVarSymbol(se->var) || isArgSymbol(se->var))) continue;
+    if (!se->var->isRef()) continue;
+
+    RefInfo* info = NULL;
+    RefInfoIter it = infoMap.find(se->var);
+    if (it == infoMap.end()) {
+      info = new RefInfo(se->var);
+      todo.push_back(se->var);
+      infoMap[se->var] = info;
+    } else {
+      info = it->second;
+    }
+
+    info->todo.push_back(se);
+  }
+
+  for_vector(Symbol, sym, todo) {
```

Comment:
Or maybe not, since you subsequently clear infoMap. I think you could remove todo here and in the comuteDotLocal case.